src/hipscat_import/soap/resume_plan.py

```diff
@@ -0,0 +1,155 @@
+
+    COUNTING_STAGE = "counting"
+    SOURCE_MAP_FILE = "source_object_map.npz"
+
+    def __init__(self, args: SoapArguments):
+        if not args.tmp_path:  # pragma: no cover (not reachable, but required for mypy)
+            raise ValueError("tmp_path is required")
+        super().__init__(resume=args.resume, progress_bar=args.progress_bar, tmp_path=args.tmp_path)
+        self.gather_plan(args)
+
+    def gather_plan(self, args):
+        """Initialize the plan."""
+        with tqdm(total=5, desc="Planning ", disable=not self.progress_bar) as step_progress:
+            ## Make sure it's safe to use existing resume state.
+            super().safe_to_resume()
+            step_progress.update(1)
+            self.count_keys = []
+
+            ## Validate existing resume state.
+            if self.is_counting_done():
+                return
+            step_progress.update(1)
+
+            source_map_file = file_io.append_paths_to_pointer(self.tmp_path, self.SOURCE_MAP_FILE)
+            if file_io.does_file_or_directory_exist(source_map_file):
+                source_pixel_map = np.load(source_map_file, allow_pickle=True)["arr_0"].item()
+            else:
+                source_pixel_map = source_to_object_map(args)
+                np.savez_compressed(source_map_file, source_pixel_map)
+            self._set_sources_to_count(source_pixel_map)
+            step_progress.update(1)
+
+    def wait_for_counting(self, futures):
+        """Wait for counting stage futures to complete."""
+        self.wait_for_futures(futures, self.COUNTING_STAGE)
+
+    def is_counting_done(self) -> bool:
+        """Are there sources left to count?"""
+        return self.done_file_exists(self.COUNTING_STAGE)
+
+    def _set_sources_to_count(self, source_pixel_map):
+        """Fetch a triple for each source pixel to join and count.
+
+        Triple contains:
+            - source pixel
+            - object pixels (healpix pixel with both order and pixel, for aligning and
+              neighboring object pixels)
+            - source key (string of source order+pixel)
+
+        """
+        counted_keys = set(self.read_log_keys(self.COUNTING_STAGE))
+        self.count_keys = [
+            (hp_pixel, object_pixels, f"{hp_pixel.order}_{hp_pixel.pixel}")
+            for hp_pixel, object_pixels in source_pixel_map.items()
+            if f"{hp_pixel.order}_{hp_pixel.pixel}" not in counted_keys
+        ]
+
+
+def source_to_object_map(args):
+    """Build a map of (source order/pixel) to the (object order/pixel)
+    that are aligned, as well as neighboring object pixels.
+    """
+    object_catalog = Catalog.read_from_hipscat(args.object_catalog_dir)
+    source_catalog = Catalog.read_from_hipscat(args.source_catalog_dir)
+
+    ## Direct aligment from source to object
+    ###############################################
+    grouped_alignment = align_trees(
+        object_catalog.pixel_tree, source_catalog.pixel_tree, "outer"
```

Comment:
I know it probably shouldn't happen for a source and object catalog, but since it's an outer alignment what happens  if one of the pixels is None?